server/src/main/java/org/opensearch/indices/replication/RemoteStoreReplicationSource.java

```diff
@@ -0,0 +1,101 @@
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/**
+ * Implementation of a {@link SegmentReplicationSource} where the source is remote store.
+ *
+ * @opensearch.internal
+ */
+public class RemoteStoreReplicationSource implements SegmentReplicationSource {
+
+    private static final Logger logger = LogManager.getLogger(PrimaryShardReplicationSource.class);
+
+    private final IndexShard indexShard;
+
+    public RemoteStoreReplicationSource(IndexShard indexShard) {
+        this.indexShard = indexShard;
+    }
+
+    @Override
+    public void getCheckpointMetadata(
+        long replicationId,
+        ReplicationCheckpoint checkpoint,
+        ActionListener<CheckpointInfoResponse> listener
+    ) {
+        FilterDirectory remoteStoreDirectory = (FilterDirectory) indexShard.remoteStore().directory();
+        FilterDirectory byteSizeCachingStoreDirectory = (FilterDirectory) remoteStoreDirectory.getDelegate();
+        RemoteSegmentStoreDirectory remoteDirectory = (RemoteSegmentStoreDirectory) byteSizeCachingStoreDirectory.getDelegate();
+
+        Map<String, StoreFileMetadata> metadataMap;
+        // TODO: Need to figure out a way to pass this information for segment metadata via remote store.
+        final Version version = indexShard.getSegmentInfosSnapshot().get().getCommitLuceneVersion();
+        try {
+            metadataMap = remoteDirectory.readLatestMetadataFile()
+                .entrySet()
+                .stream()
+                .collect(
+                    Collectors.toMap(
+                        e -> e.getKey(),
+                        e -> new StoreFileMetadata(
+                            e.getValue().getOriginalFilename(),
+                            e.getValue().getLength(),
+                            Store.digestToString(Long.valueOf(e.getValue().getChecksum())),
+                            version,
+                            null
+                        )
+                    )
+                );
+            // TODO: GET current checkpoint from remote store.
+            listener.onResponse(new CheckpointInfoResponse(checkpoint, metadataMap, null));
+        } catch (IOException e) {
+            logger.error("Error fetching checkpoint metadata from remote store", e);
+            listener.onFailure(e);
+        }
+    }
+
+    @Override
+    public void getSegmentFiles(
+        long replicationId,
+        ReplicationCheckpoint checkpoint,
+        List<StoreFileMetadata> filesToFetch,
+        IndexShard indexShard,
+        ActionListener<GetSegmentFilesResponse> listener
+    ) {
+        try {
+            indexShard.syncSegmentsFromRemoteSegmentStore(false, true, false);
+            listener.onResponse(new GetSegmentFilesResponse(Collections.emptyList()));
+        } catch (Exception e) {
+            logger.error("Failed to sync segments", e);
```

Comment:
Same comment as above. Will this result in double logging?

Reply:
Nope. We're not logging the error anywhere in the call stack. Eventually we're just wrapping this exception with `ReplicationFailedException` [here](https://github.com/opensearch-project/OpenSearch/blob/1e08b5a075fa6018be6c0af959b2d638c2743d56/server/src/main/java/org/opensearch/indices/replication/SegmentReplicationTargetService.java#L395)